sdk/messaging/azservicebus/receiver.go

```diff
@@ -341,123 +353,128 @@ func (r *Receiver) receiveMessage(ctx context.Context, options *ReceiveMessagesO
+		defer cancel()
+
 		// start the drain asynchronously. Note that we ignore the user's context at this point
 		// since draining makes sure we don't get messages when nobody is receiving.
-		if err := linksWithID.Receiver.DrainCredit(context.Background()); err != nil {
-			if err := r.amqpLinks.RecoverIfNeeded(context.Background(), linksWithID.ID, err); err != nil {
-				log.Writef(internal.EventReceiver, "Failed to recover links after a failed drain: %s", err.Error())
-				return ret(err)
-			}
-
-			// ignore the drain error - the link will be reopened at the next chance anyways
-			// and there's nothing the user is expected to or needs to do here.
-			return ret(nil)
+		if err := linksWithID.Receiver.DrainCredit(ctx); err != nil {
+			// note that cancelling a DrainCredit means we're in an indeterminate state
+			// so we treat that as a "must recover" error.
+			return flushAndReturn(err)
 		}
 
-		// Draining data from the receiver's prefetched queue. This won't wait for new messages to
-		// arrive, so it'll only receive messages that arrived prior to the drain.
-		for {
-			am, err := linksWithID.Receiver.Prefetched(context.Background())
+		return flushAndReturn(nil)
+	}
 
-			if am == nil || internal.IsCancelError(err) {
-				return all, nil
-			}
+	// we only get here if we got exactly the # of messages they asked for
+	// in the initial set of link.Receive() calls above.
+	return all, nil
+}
 
-			if err != nil {
-				return ret(err)
-			}
+func fetchMessages(ctx context.Context, receiver internal.AMQPReceiver, maxMessages int, defaultTimeAfterFirstMessage time.Duration, messages *[]*ReceivedMessage) error {
+	if err := receiver.IssueCredit(uint32(maxMessages)); err != nil {
+		return err
+	}
+
+	var cancel context.CancelFunc
 
-			all = append(all, newReceivedMessage(am))
+	for {
+		amqpMessage, err := receiver.Receive(ctx)
+
+		if err != nil {
+			return err
+		}
+
+		*messages = append(*messages, newReceivedMessage(amqpMessage))
+
+		if len(*messages) == maxMessages {
+			return nil
+		}
+
+		if cancel == nil {
+			// replace the context that we're using for everything with a new one that will cancel
+			// after a period of time.
+			ctx, cancel = context.WithTimeout(ctx, defaultTimeAfterFirstMessage)
+			defer cancel()
 		}
 	}
+}
 
-	return all, nil
+// flushPrefetchedMessages makes a best-effort attempt at draining any messages on the link. If the link is dead,
+// or times out when draining we will close it quickly. The next operation on the link will
+// recover it.
+func flushPrefetchedMessages(ctx context.Context, receiver internal.AMQPReceiver, messages *[]*ReceivedMessage) {
+	// Draining data from the receiver's prefetched queue. This won't wait for new messages to
+	// arrive, so it'll only receive messages that arrived prior to the drain.
+	for {
+		am, err := receiver.Prefetched(ctx)
+
+		// we've removed any code of consequence from Prefetched.
+		if am == nil || err != nil || internal.IsCancelError(err) {
```

Comment:
I don't think `internal.IsCancelError(err)` clause will ever be hit when `err` isn't `nil`.

Reply:
:facepalm:

Reply:
Fixed...:|